python/source.py

```diff
@@ -875,34 +876,74 @@ def get_fields(self, sim):
         )
         waist_xy = np.array([np.real(X0[0]), np.real(X0[1])])[:, np.newaxis, np.newaxis]
 
+        # Find large imaginary argument hankel shift
+        r, _ = self.get_r_rhat(X, X0)
+        kr = (k * r).flatten()
+        max_imag = np.abs(kr.imag[np.argmax(np.abs(kr.imag))])
+        max_float = int(np.log(float_info.max) / 2)
+        shift = 0 if np.abs(max_imag) < max_float else max_imag - max_float
```

Comment:
I use `sys.float_info.max` to create a threshold. If the argument into the hankel functions ($kr$) has an imaginary component that is bigger (in magnitude) than this threshold, then I scale my entire system by a factor ($e^{-s}$ for scale s), to bring the magnitude smaller than what will overflow in the exponential. Without this extra `scale` component, reformulating with the exponential bessel functions would be useless because the overflow would instead still happen in my normalizing exponential term $e^{\pm kr}$ rather than in the old hankel functions.

If we want to use another way of deciding the threshold besides `sys.float_info.max` then I can change it.